qcore/decorators.py

```diff
@@ -63,23 +64,35 @@ def __str__(self):
     def __repr__(self):
         return self.__str__()
 
-    def __cmp__(self, other):
-        """Compare objects for equality (so we can run tests that do an equality check).
-
-        This is implemented using __cmp__ because this is the only way that works in both
-        Cythonized and non-Cythonized classes.
-
-        """
-        if self.__class__ is other.__class__ and self.decorator == other.decorator and \
-                self.instance == other.instance:
-            return 0
-        else:
-            return -1
+    # This awkward implementation is necessary so that binders can be compared for equality across
+    # Cythonized and non-Cythonized Python 2 and 3. In pure-Python compiled classes, Cython only
+    # supports overriding __richcmp__, not __eq__ (https://github.com/cython/cython/issues/690),
+    # but if __eq__ is defined it throws an error.
+    if cython.compiled:
+        def __richcmp__(self, other, op):
+            """Compare objects for equality (so we can run tests that do an equality check)."""
+            if op in (2, 3): # ==, !=
+                if self.__class__ is other.__class__ and self.decorator == other.decorator and \
+                        self.instance == other.instance:
+                    equal = True
+                else:
+                    equal = False
+                return equal if op == 2 else not equal
+            else:
+                return NotImplemented
 
     def __hash__(self):
         return hash(self.decorator) ^ hash(self.instance)
 
 
+if not cython.compiled:
+    def __eq__(self, other):
+        return self.__class__ is other.__class__ and self.decorator == other.decorator and \
+            self.instance == other.instance
+    DecoratorBinder.__eq__ = __eq__
+    del __eq__
+
```

Comment:
It doesn't work:

  File "qcore/decorators.py", line 94, in init qcore.decorators (qcore/decorators.c:10279)
    DecoratorBinder.__richcmp__ = __richcmp__
TypeError: can't set attributes of built-in/extension type 'qcore.decorators.DecoratorBinder'

Putting __eq__ inside the class doesn't work because Cython errors out if you define __eq__ on an extension class. I tried creating __eq__ with exec() too, but that made Cython produce C code that didn't compile (probably due to an unrelated bug).